lldb/source/Target/Language.cpp

```diff
@@ -574,3 +576,164 @@ bool SourceLanguage::IsObjC() const {
+/// number is `line_number`. If no such entry is found, returns the entry
+/// closest to but after `line_number`.
+static std::optional<Address> FindAddressForLineNumber(Function &function,
+                                                       uint32_t line_number) {
+  CompileUnit *cu = function.GetCompileUnit();
+  LineTable *line_table = cu ? cu->GetLineTable() : nullptr;
+  if (line_table == nullptr)
+    return std::nullopt;
+
+  // Get the first line entry for this function.
+  AddressRange func_range = function.GetAddressRange();
+  uint32_t first_entry_idx;
+  {
+    LineEntry first_line_entry;
+    line_table->FindLineEntryByAddress(func_range.GetBaseAddress(),
+                                       first_line_entry, &first_entry_idx);
+  }
+
+  LineEntry best_match;
+  for (uint32_t entry_idx = first_entry_idx; entry_idx < line_table->GetSize();
+       entry_idx++) {
+    LineEntry next_line;
+    line_table->GetLineEntryAtIndex(entry_idx, next_line);
+
+    // Stop if this entry is outside the range of `function`.
+    Address base_addr = next_line.range.GetBaseAddress();
+    if (!func_range.ContainsFileAddress(base_addr))
+      break;
+
+    // Stop on an exact match.
+    if (next_line.line == line_number) {
+      best_match = next_line;
+      break;
+    }
+
+    // Otherwise, keep track of the best match so far.
+    if (next_line.line > line_number && next_line.line < best_match.line)
+      best_match = next_line;
+  }
+
+  return best_match.range.GetBaseAddress();
+}
+
+/// Given a list of functions, returns a map: Function -> VariableList
+/// containing local variables of each function.
+static std::map<Function *, VariableList>
+GetFuncToLocalVariablesMap(llvm::ArrayRef<Function *> funcs) {
+  std::map<Function *, VariableList> map;
+  for (Function *function : funcs) {
+    VariableList &variable_list = map[function];
+    Block &block = function->GetBlock(true /*can_create=*/);
+    block.AppendBlockVariables(
+        true /*can_create=*/, true /*get_child_block_variables=*/,
+        true /*stop_if_child_block_is_inlined_function=*/,
+        [](Variable *v) { return true; }, &variable_list);
+  }
+  return map;
+}
+
+/// Returns the first line associated with `function`.
+static uint32_t GetLineNumberForFunction(Function &function) {
+  FileSpec filespec;
+  uint32_t line_num = 0;
+  function.GetStartLineSourceInfo(filespec, line_num);
+  return line_num;
+}
+
+/// Checks if `var` is in scope inside `function` at line `line_number`.
+/// If this check can't be done, a best-effort comparison of:
+///    line_number >= var.line_number
+/// is performed.
+static bool IsVariableInScopeAtLine(Function &function, uint32_t line_number,
+                                    Variable &var) {
+  auto fallback_line_comp = [&] {
+    return line_number >= var.GetDeclaration().GetLine();
+  };
+
+  std::optional<Address> line_addr =
+      FindAddressForLineNumber(function, line_number);
```

Comment:
I should move this to the call site, so that it can be cached across iterations of the for loop